sw/device/lib/base/hardened_memory.c

```diff
@@ -0,0 +1,180 @@
+
+    // Prevent the compiler from reordering the loop; this ensures a
+    // happens-before among indices consistent with `order`.
+    barrierw(idx);
+
+    // Compute putative offsets into `src`, `dest`, and `decoys`. Some of these
+    // may go off the end of `src` and `dest`, but they will not be cast to
+    // pointers in that case. (Note that casting out-of-range addresses to
+    // pointers is UB.)
+    uintptr_t srcp = src_addr + idx;
+    uintptr_t destp = dest_addr + idx;
+    uintptr_t decoy1 = decoy_addr + (idx & 7);
+    uintptr_t decoy2 = decoy_addr + ((idx + 4) & 7);
+
+    // Branchlessly select whether to do a "real" copy or a decoy copy,
+    // depending on whether we've gone off the end of the array or not.
+    //
+    // Pretty much everything needs to be laundered: we need to launder `idx`
+    // for obvious reasons, and we need to launder the result of the select, so
+    // that the compiler cannot delete the resulting loads and stores. This is
+    // similar to having used `volatile uint32_t *`.
+    void *src = (void *)launderw(
+        ct_cmovw(ct_sltuw(launderw(idx), word_len), srcp, decoy1));
+    void *dest = (void *)launderw(
+        ct_cmovw(ct_sltuw(launderw(idx), word_len), destp, decoy2));
+
+    // Perform the copy, without performing a typed dereference operation.
+    write_32(read_32(src), dest);
+  }
+
+  // No need to launder here; it's the final use of `count`, which has been
+  // laundered above. This ensures that we have traversed the entire random
+  // order sequence.
+  HARDENED_CHECK_EQ(count, expected_count);
+}
+
+// The source of randomness for shred, which may be replaced at link-time.
+OT_ATTR_WEAK
+uint32_t hardened_memshred_random_word(void) { return 0xcaffe17e; }
+
+void hardened_memshred(uint32_t *dest, size_t word_len) {
+  random_order_t order;
+  random_order_init(&order, word_len);
+
+  size_t count = 0;
+  size_t expected_count = random_order_len(&order);
+
+  uintptr_t data_addr = (uintptr_t)dest;
+
+  uint32_t decoys[8];
+  uintptr_t decoy_addr = (uintptr_t)&decoys;
+
+  for (; count < expected_count; count = launderw(count) + 1) {
+    size_t idx = launderw(random_order_advance(&order)) * 4;
+    barrierw(idx);
+
+    uintptr_t datap = data_addr + idx;
+    uintptr_t decoy = decoy_addr + (idx & 7);
+
+    void *data = (void *)launderw(
+        ct_cmovw(ct_sltuw(launderw(idx), word_len), datap, decoy));
+
+    // Write a freshly-generated random word to `*data`.
+    write_32(hardened_memshred_random_word(), data);
+  }
+
+  HARDENED_CHECK_EQ(count, expected_count);
+}
+
+hardened_bool_t hardened_memeq(const uint32_t *lhs, const uint32_t *rhs,
+                               size_t word_len) {
+  random_order_t order;
+  random_order_init(&order, word_len);
+
+  size_t count = 0;
+  size_t expected_count = random_order_len(&order);
+
+  uintptr_t a_addr = (uintptr_t)lhs;
+  uintptr_t b_addr = (uintptr_t)rhs;
```

Comment:
nit: {l,r}hs_addr for consistency? (also in the loop if you agree).
```suggestion
  uintptr_t lhs_addr = (uintptr_t)lhs;
  uintptr_t lhs_addr = (uintptr_t)rhs;
```